### PR TITLE
Revise kernel #includes

### PR DESCRIPTION
**Pull request by fingolfin**
This PR revises all kernel #includes, stripping them down to the (for now) minimum. Key ideas and benefits:

1. each header file should #include any other header files that need to be #included before it; at the very least, `system.h`, which defines core types and includes `config.h`; this way, if you need a header, you can just #include it w/o worrying what else it in turn needs.
2. each C file should only include headers it really needs; this keeps compile times to a minimum when changing between revisions and branches, when editing a header etc.
3. inside a C file, the #includes are sorted alphabetically, to avoid having to think were one should put a new one; some exceptions:
  * I put all the `hpc/` includes to the end
   * actually, system headers come at the *very end*
   * the header matching the current C file comes first (so `#include <src/foo.h>` is first in `src/foo.c`). The idea here being that this way, each header is included first at least once, which helps to ensure goal 1 keeps being satisfied
4. No comments after each include what it is about. This is documented inside the headers, and could change, and it's just too much work to maintain; and I have yet to find any benefit in it.

Some TODOs for the future, after this PR has been merged:
* perhaps change the "central" header file from `system.h` to something else. 
* split `system.h` into the part providing essential types, defines and includes; and the part defining tons of variabls
* split `gap.h` (say into an error.h, plus more); indeed, I have some commits for that already queued up, but didn't want to mix them into this PR, which is already quite big
*  various FIXMEs I added to #includes should be resolved
* I'd like to clean up the file headers (the comments at the top of almost every .c and .h file), to a uniform format, which states a sane copyright, and probably drops pointless things (like the name of the file, which is useless, and gets out-of-sync if the file is renamed; or the "author" of the file, which does not reflect reality for almost all of those files)

### CODE REVIEW

**Comment by fingolfin** [comment]
Oh, and please don't get too upset about the commit "Move typedef ObjFunc to system.h". It's simple pragmatic, just like pushing the typedefs for `Stat` and `Expr` into `system.h`; but it need not stay like that forever. But for now, it simply made it much easier to streamline the includes. Once things have settled a bit, we can look into moving those typedefs into "better" places, whatever that means.

**Comment by ChrisJefferson** [comment]
Just a check -- did you make sure this didn't break any package?

File: src/calls.h [hunk 1]
```diff
@@ -45,10 +45,8 @@
 #ifndef GAP_CALLS_H
 #define GAP_CALLS_H
 
-#include <src/gap.h>
-#include <src/gaputils.h>
-#include <src/lists.h>
-#include <src/integer.h>
+#include <src/gaputils.h>    // ARRAY_SIZE
```
Comment (ChrisJefferson) [inline src/calls.h]: Maybe remove this comment, if we aren't having them anywhere else?

**Comment by fingolfin** [comment]
Why / how could this possibly break packages? Almost all of them include only compiled.h, but even if not: this change ensures each header becomes independently usable, by including other headers one *must* include before it (if not, a compile error ensues).

Besides, Travis includes a test compiling most deposited packages, and that test passed.

File: src/gap.c [hunk 2]
```diff
@@ -10,110 +10,40 @@
...
+#include <src/gvars.h>
+#include <src/integer.h>
+#include <src/lists.h>
+#include <src/opers.h>
+#include <src/plist.h>
+#include <src/precord.h>
+#include <src/records.h>
+#include <src/read.h>
+#include <src/saveload.h>
+#include <src/stats.h>    // for ClearError
```
Comment (ChrisJefferson) [inline src/gap.c]: Also here, why a comment?

File: src/streams.c [hunk 3]
```diff
@@ -12,49 +12,27 @@
...
+#include <src/gap.h>
+#include <src/gapstate.h>
+#include <src/gvars.h>
+#include <src/lists.h>
+#include <src/io.h>
+#include <src/plist.h>
+#include <src/precord.h>
+#include <src/read.h>
+#include <src/records.h>
+#include <src/stats.h>      // for ClearError
```
Comment (ChrisJefferson) [inline src/streams.c]: another comment

**Comment by ChrisJefferson** [comment]
This is fine, I have some minor comments on some random looking comments (did you add these while sorting out the headers?)

File: src/calls.h (hunk 1 above)
Comment (fingolfin) [reply src/calls.h]: Huh, indeed, not sure how that survived my script.

File: src/gap.c (hunk 2 above)
Comment (fingolfin) [reply src/gap.c]: Should be: ˋFIXME: only for ClearErrorˋ. Point is, ClearError really should be somewhere else, several files include stats.h just for it.

File: src/calls.h (hunk 1 above)
Comment (fingolfin) [reply src/calls.h]: (Or maybe I added it as annotation? Anyway, will remove)

**Comment by fingolfin** [comment]
@ChrisJefferson updated

**Pull request merged**